### ISSUE TITLE
vars in nested blocks not overwrite vars in upper blocks

### ISSUE THREAD

**Issue by Fmajor** [ansible/ansible#76883]
### Summary

vars in nested blocks not overwrite vars in upper blocks, see the below example.

### Issue Type

Bug Report

### Component Name

none

### Ansible Version

```console
$ ansible --version
2.11.3
```

### Configuration

```console
$ ansible-config dump --only-changed
# This bug has nothing to do with ansible configs
```

### OS / Environment

Gentoo

### Steps to Reproduce

```yaml (paste below)
---
- hosts: localhost
  gather_facts: False
  vars:
    data0: -1
  tasks:
    - name: level0
      vars:
        data0: 0
      block:
        - name: level1
          vars:
            data0: 1
          block:
            - name: level2
              vars:
                data0: 2
              block:
                - vars:
                    data0: local
                  debug_var: # this is my custom action, it will just print the value of `var`
                    var:
                      - data0
                - debug_var:
                    var:
                      - data0
            - debug_var:
                var:
                  - data0
        - debug_var:
            var:
              - data0
```

### Expected Results

```yaml 
# get this output (I have a custom callback, it will print the nested block structure)
# I expect the `vars` in deeper block level will overwrite upper vars with the same name
{"hosts": ["localhost"]}
                  > level0
                    > level1
                      > level2
[@localhost][  1]       * <debug_var> +0.0s +0.0s [debug_var]  SUCCESS
  {
    "data0": "local"
  }
[@localhost][  2]       * <debug_var> +0.0s +0.1s [debug_var]  SUCCESS
  {
    "data0": 2
  }
[@localhost][  3]     * <debug_var> +0.0s +0.1s [debug_var]  SUCCESS
  {
    "data0": 1
  }
[@localhost][  4]   * <debug_var> +0.0s +0.1s [debug_var]  SUCCESS
  {
    "data0": 0
  }
[@localhost][  5] * <debug_var> +0.0s +0.1s [debug_var]  SUCCESS
  {
    "data0": -1
  }
```

### Actual Results

```console
# get this output (I have a custom callback, it will print the nested block structure)
# but only the vars in level0 take effect.
{"hosts": ["localhost"]}
                  > level0
                    > level1
                      > level2
[@localhost][  1]       * <debug_var> +0.0s +0.0s [debug_var]  SUCCESS
  {
    "data0": "local"
  }
[@localhost][  2]       * <debug_var> +0.0s +0.1s [debug_var]  SUCCESS
  {
    "data0": 0
  }
[@localhost][  3]     * <debug_var> +0.0s +0.1s [debug_var]  SUCCESS
  {
    "data0": 0
  }
[@localhost][  4]   * <debug_var> +0.0s +0.1s [debug_var]  SUCCESS
  {
    "data0": 0
  }
[@localhost][  5] * <debug_var> +0.0s +0.1s [debug_var]  SUCCESS
  {
    "data0": -1
  }
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by Fmajor** [comment]
for the above test, I find that 
* the `vars` in blocks will overwrite `vars` in playbook.
* the `vars` in single task (debug_var in this example) will overwrite `vars` in its parent block
* the `vars` in block will NOT overwrite `vars` in its parent block
the last result is confusing for me

**Comment by mkrizek** [comment]
Smaller reproducer that does not use custom plugins:
```yaml
- hosts: localhost                                                                                                                                                                                                                                                                                                    
  gather_facts: false
  vars:
    foo: -1
  tasks:
    - debug:
        var: foo
    - block:
      - debug:
          var: foo
      - block:
        - debug:
            var: foo
        vars:
          foo: 1
      vars:
        foo: 0
```

Before #76888:
```
PLAY [localhost] *****************************************************************************************************************************************************************************************************************************************************************************************************

TASK [debug] *********************************************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "foo": -1
}

TASK [debug] *********************************************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "foo": 0
}

TASK [debug] *********************************************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "foo": 0
}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

After #76888:
```
PLAY [localhost] *****************************************************************************************************************************************************************************************************************************************************************************************************

TASK [debug] *********************************************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "foo": -1
}

TASK [debug] *********************************************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "foo": 0
}

TASK [debug] *********************************************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "foo": 1
}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

**Comment by sivel** [comment]
Duplicate of https://github.com/ansible/ansible/issues/75286

**Comment by bcoca** [comment]
closing as dupe
